File: helloworld.cpp
// Configuration of C++. Need following 2 lines
#include <iostream>
#include <cmath>

using namespace std;

// // 1st way: functions creating before calling it
// void sayHi(string name, int age) {
//   cout << "Hello, " << name << endl;
//   cout << "You are " << age << " years old." << endl;
// }

// 2nd: Create an abstract and complete it below.
void sayHi(string name, int age);
int getMax(int num1, int num2, int num3);
string getDayOfWeek(int dayNum, int startSunday);
int power(int base, int expo);

// A little differ to Java
class Book {
  public:
    string title;
    string author;
    int pages;

    // Constructor functions
    // This is called whenever we create a book object
    Book(string aTitle, string aAuthor, int aPages) {
      cout << "Creating Object..." << endl;
      title = aTitle;
      author = aAuthor;
      pages = aPages;
    }

    // We can also have multiple constructor objects
    Book(string aTitle, string aAuthor) {
      cout << "Creating Object w/o pages..." << endl;
      title = aTitle;
      author = aAuthor;
    }
    Book(string aTitle) {
      cout << "Creating Object w/o author and pages..." << endl;
      title = aTitle;
    }
};

class Student {
  public:
    string name;
    string major;
    double gpa;
    Student (string aName, string aMajor, double aGpa) {
      name = aName;
      major = aMajor;
      gpa = aGpa;
    }
};

int main() {
  // cout: console out
  // endl: end line

  // cout << "Hello, world!" << endl;
  // cout << "My name is Richard." << endl;

  // Drawing a shape
  // cout << "   /｜" << endl;
  // cout << "  / ｜" << endl;
  // cout << " /  ｜" << endl;
  // cout << "/___｜" << endl;

  // // Creating variables
  // string characterName = "Richard";
  // int characterAge = 20;
  // cout << "My name is " << characterName << "." << endl;
  // cout << "I am " << characterAge << " years old." << endl;

  // // Data Types
  // char grade = 'A';
  // string phrase = "c++ is fun";
  // int age = 20;
  // float avg = 90.1
  // double gpa = -4.5 // can store MORE decimal points than float
  // boolean isMale = True;

  // // C++ endl logic
  // cout << "Hello, world!"; // did not start a new line
  // cout << "Hello, world! \n"; // \n will start a new line
  // cout << "My name is Richard. \n" << endl;

  // // String functions
  // string phrase = "Richard";
  // cout << phrase.length() << endl;
  // cout << phrase[0] << endl;
  // phrase[0] = 'B';
  // cout << phrase[0] << endl;
  // // find(str or char, find arg after int elem)
  // cout << phrase.find("ard", 0) << endl; // returns 4
  // cout << phrase.find("ard", 4) << endl; // returns 4
  // cout << phrase.find("ard", 5) << endl; // returns inf (DNE)
  // // substr(int start index, int length)
  // string phrasesub = phrase.substr(1, 3);
  // cout << phrasesub << endl; // returns ich

  // // Basic C++ mathematics
  // cout << 5 + 7 * 3 % 2 << endl; // prints 6
  // cout << 29 / 10 << endl; // prints 2 (floor div between ints)
  // cout << 5. / 2 << endl; // prints 2.5
  // double dnum = 3.3;
  // dnum *= 2;
  // cout << dnum << endl; // 6.6

  // // Math functions relies <cmath>
  // cout << pow(2, 5) << endl; // 32, 2 to power of 5
  // cout << sqrt(36) << endl; // 6
  // cout << round(4.5) << endl; // 5
  // cout << floor(4.999999999) << endl; // 4
  // cout << fmax(3, 10) << endl; // return the larger elem (10)
  // cout << fmin(3, 10) << endl; // 3. Can only take 2 elems.

  // // Prompting
  // int age; // if char age, will return first char of input
  // cout << "Enter your age: ";
  // cin >> age;
  //
  // cout << "You are " << age << " years old." << endl;
  //
  // string name; // if char age, will return first char of input
  // cout << "Enter your name: ";
  // getline(cin, name);
  //
  // cout << "Hello, " << name << "." << endl;

  // // Build a calculator (user input)
  // double num1, num2; // declare variables easily
  // cout << "Enter first number: ";
  // cin >> num1;
  // cout << "Enter second number: ";
  // cin >> num2;
  //
  // cout << num1 + num2 << endl;

  // // Build a mad libs game
  // string color, pluralNoun, celebrity;
  //
  // cout << "Enter a color: ";
  // getline(cin, color);
  // cout << "Enter a plural noun: ";
  // getline(cin, pluralNoun);
  // cout << "Enter a celebrity: ";
  // getline(cin, celebrity);
  //
  // cout << "Roses are " << color << endl;
  // cout << pluralNoun << " are blue" << endl;
  // cout << "I love " << celebrity << endl;

  // // Arrays
  // // different with java, we put [] after var name.
  // int luckyNums[] = {7, 6, 49, 24}; // use {} to initial elems
  // cout << luckyNums[3] << endl; // 24
  // luckyNums[3] = 9;
  // cout << luckyNums[3] << endl; // 9
  //
  // int nums[20];
  // cout << nums[3] << endl; // 0
  // nums[3] = 9;
  // cout << nums[3] << endl; // 9

  // // Functions
  // cout << "Before \n";
  // sayHi("Richard", 60);
  // sayHi("Jason", 32);
  // sayHi("Abby", 48);
  // cout << "After \n";

  // // If statement
  // bool isMale = true; // lowercase true
  // bool isTall = false;
  //
  // // logic similar to java
  // if(isMale && isTall) {
  //   cout << "You are a tall male." << endl;
  // }
  // else if (isMale && !isTall) {
  //   cout << "You are a short male." << endl;
  // }
  // else if (!isMale && isTall) {
  //   cout << "You are tall but not male." << endl;
  // } else {
  //   cout << "You are neither tall nor male." << endl;
  // }

  // // If statement using comparisons
  // cout << getMax(3, 5, -8) << endl;

  // // Building a better calculator
  // double num1, num2, result;
  // char op;
  //
  // cout << "Enter first number: ";
  // cin >> num1;
  // cout << "Enter operator: ";
  // cin >> op;
  // cout << "Enter second number: ";
  // cin >> num2;
  //
  // if (op == '+') {
  //   result = num1 + num2;
  //   cout << result << endl;
  // } else if (op == '-') {
  //   result = num1 - num2;
  //   cout << result << endl;
  // } else if (op == '*') {
  //   result = num1 * num2;
  //   cout << result << endl;
  // } else if (op == '/') {
  //   result = num1 / num2;
  //   cout << result << endl;
  // } else {
  //   cout << "Invalid operation." << endl;
  // }

  // // Switch statements
  // cout << getDayOfWeek(28, 7) << endl;

  // // While loops
  // // check condition BEFORE execute
  // int index = 5;
  // while(index < 5) {
  //   cout << "WHILE" << endl;
  //   index++;
  // }
  // // check condition AFTER execute
  // do {
  //   cout << "DO" << endl;
  //   index++;
  // } while(index < 5);

  // // While Loops - Build a Guessing Game
  // int secretNum = 7;
  // int guess;
  //
  // while(guess != secretNum) {
  //   cout << "Enter guess: " << endl;
  //   cin >> guess;
  //   if (guess < secretNum) {
  //     cout << "Make a larger guess!" << endl;
  //   }
  //   if (guess > secretNum) {
  //     cout << "Make a smaller guess!" << endl;
  //   }
  // }
  //
  // cout << "You Win!" << endl;

  // // For Loops
  // for (int i = 1; i < 3; i++) {
  //   cout << i << endl;
  // }

  // // Exponent Functions
  // cout << power(2, 0) << endl;
  // cout << power(2, 3) << endl;

  // // 2d arrays and nested for Loops
  // // The numbers in [] is pretty important
  // int numberGrid[3][2] = {
  //                           {1, 2},
  //                           {3, 4},
  //                           {5, 6}
  //                       };
  //
  // // only prints location in memory
  // cout << numberGrid << endl;
  //
  // cout << numberGrid[1][0] << endl; // 3
  //
  // for (int i = 0; i < 5; i++) {
  //   for (int j = i; j < 5; j++) {
  //     cout << '*';
  //   }
  //   cout << '\n';
  // }

  // // Pointers
  // int age = 20;
  // double gpa = 0.1;
  // string name = "Richard";
  //
  // // the address will change every time
  // // we call &age a POINTER
  // cout << "Age: " << &age << endl;
  // cout << "GPA: " << &gpa << endl;
  // cout << "Name:" << &name << endl;
  //
  // // int pAge is illegal since &age is not int
  // int *pAge = &age;
  // cout << "pAge: " << pAge << endl; // &age
  // // de-referencing - extract the value in that pointer
  // cout << "*pAge: " << *pAge << endl; // age (20)
  // cout << "*&age: " << *&age << endl; // age (20)

  // // Classes and objects
  // // Class is actually a new data type created artificially
  // // Object is the instance of a class
  // Book book1;
  // book1.title = "Harry Potter";
  // book1.author = "JK Rowling";
  // book1.pages = 500;
  //
  // cout << book1.title << endl;
  //
  // Book book2;
  // book2.title = "Lord of the Rings";
  // book2.author = "Tolkeing";
  // book2.pages = 700;
  //
  // cout << book2.title << endl;

  // // Constructor Functions
  // // Can initialize attributes wile creating object
  // // instead of manually setting up
  // Book book1("Harry Potter", "JK Rowling", 500);
  // cout << book1.title << endl;
  //
  // Book book2("Lord of the Ring", "Tolkeing", 700);
  // cout << book2.title << endl;

  // Object functions
  Student student1("Jim", "Business", 2.4);
  Student student2("Pam", "Art", 3.6);

  

  return 0;
}

// int getMax(int num1, int num2, int num3) {
//   if (num1 >= num2 && num1 >= num3) {
//     return num1;
//   } else if (num2 >= num1 && num2 >= num3) {
//     return num2;
//   } else {
//     return num3;
//   }
// }

// // 2nd
// void sayHi(string name, int age) {
//   cout << "Hello, " << name << endl;
//   cout << "You are " << age << " years old." << endl;
// }

// string getDayOfWeek(int dayNum, int startSunday) {
//   string dayName;
//
//   // // We need to copy and paste 7 times if use if statement
//   // if(dayNum == 0) {
//   //   dayName = "Sunday";
//   // } // else if == 1 ...
//
//   // Switch statements
//   switch(dayNum % startSunday) {
//   case 0:
//     dayName = "Sunday";
//     break;
//   case 1:
//     dayName = "Monday";
//     break;
//   case 2:
//     dayName = "Tuesday";
//     break;
//   case 3:
//     dayName = "Wednesday";
//     break;
//   case 4:
//     dayName = "Thursday";
//     break;
//   case 5:
//     dayName = "Friday";
//     break;
//   case 6:
//     dayName = "Saturday";
//     break;
//   // else case
//   default:
//     dayName = "Invalid day name";
//   }
//
//   return dayName;
// }

// int power(int base, int expo) {
//   int res = 1;
//   for (int i = 0; i < expo; i++) {
//     res = res * base;
//   }
//   return res;
// }

File: ruby/helloworld.rb
# # 'print' no skipping line
# print "Hello, print1"
# print "Hello, print2"
# # 'puts' skips a line
# puts "Hello, put1"
# puts "Hello, put2"
#
# # Draw an easy shape
# puts " /___|"
# puts "    /|"
# puts "   / |"
# puts "  /  |"

# # Variables
# character_name = "Richard"
# # int can't be inserted between strings
# character_age = "20"
# # parethesis is not necessary but recommended
# puts ("Hello, my name is " + character_name)
# puts ("I am " + character_age + " years old.")

# # Data Types
# name = "Richard" # string
# age = 20 # int
# gpa = 3.1 # float/double
# isMale = true # boolean
# flaws = nil # nil - does not have a value (ie. initializaiton only)

# # Working with strings
# # String functions:
# phrase = "   UC Berkeley is in UC system   "
# puts phrase.upcase()
# puts phrase.downcase()
# puts phrase.strip() # dealing with leading and trilling whitespaces
# puts phrase.length()
# puts phrase.include? "system" #check if substring included in phrase
# puts phrase.include? "state"
# puts phrase[4] # "C"
# puts phrase[4,3] # "C B"
# puts phrase.index("B")

# # Working with numbers
# puts -5.86543 # double negative
# puts 5 + 7
# puts 5 / 7
# puts 5.5 * 2.2
# puts 10 % 3
#
# num = -20.487
# puts num.abs()
# puts num.abs().round()
# puts num.abs().ceil()
# puts num.floor()
#
# puts Math.log(36)
# puts Math.sqrt(36)

# # Getting user input
# puts "Enter Your Name: "
# # name = gets
# # # This would print "Hello, Richard"
# # #                  ", you are cool!"
# # puts ("Hello, " + name + ", you are cool!")
# # # To avoid this, we do:
# name = gets.chomp()
# # This would print "Hello, Richard, you are cool!"
# puts ("Hello, " + name + ", you are cool!")
# puts "Enter Your Age: "
# age = gets.chomp()
# puts ("Your name is: " + name + ", and you are "+age+"years old.")

# # Build a calculator
# puts "Enter a number: "
# num1 = gets.chomp()
# puts "Enter another number: "
# num2 = gets.chomp()
# # suppose 1 = 3, 2 = 7; converts to string auto, ie. Prints "37"
# puts "The result is: "
# puts (num1 + num2)
# # 3 + 7 = 10
# # to_i or to_f (int vs float)
# # or num1 = gets.chomp().to_f
# puts (num1.to_i + num2.to_i)

# # Build a Mab Libs Game
# puts ("Enter a color: ")
# color = gets.chomp()
# puts ("Enter a plural noun: ")
# pluralNoun = gets.chomp()
# puts ("Enter a celebrity: ")
# celebrity = gets.chomp()
#
# puts ("The generated poem is: ")
# puts ("Roses are " + color)
# puts (pluralNoun + " are blue")
# puts ("I love " + celebrity)

# Arrays
# friends = Array["Kevin", "Karen", "Oscar", "Yuri"]
# # Every element occupies a line
# puts friends
# # access a specific element
# puts friends[1]
# puts friends[-1] # oscar
# # slice (start index, number of elements start from that index)
# puts friends[2,1] # Oscar
# puts friends[1, 3] # karen, \n oscar, \n yuri
# friends = Array.new
# friends[0] = 'Michael'
# friends[5] = 'Holly'
# puts friends[0]
# puts friends[5]
# puts friends[3] # nothing
# puts friends.length()
# puts friends.include? "Michael"
# puts friends.reverse()
# friends = Array["Michael", "Karen", "Kate"]
# puts friends.sort() # alphabeitically

# # Hashes - A type of data structure
# # can store (key, value) pair
# # both : and "" work
# states = {
#   :Pennsylvania => "PA",
#   "New York" => "NY",
#   "Oregon" => "OR",
#   3 => 'Number'
# }
# # Prints this structure {key1 => value1, key2 => value2, ...}
# puts states
# # Prints value by giving key
# puts states["New York"] # NY
# puts states[1] # if no such key, print an empty line
# puts states[:Pennsylvania] # PA
# puts states[3]

# # Methods (functions)
# # def - end
# def sayHi
#   puts "Hello user"
# end
# # execute: type method Name
# puts "top"
# sayHi
# puts "bottom"
# # methods with arguments, also default value
# def sayHi(name="No name", age=-1)
#   # puts ("Hello" + 1) illegal
#   puts ("Hello " + name + ", you are " + age.to_s)
    # age can't be int
# end
# sayHi # only work when there's defalt value
# sayHi("Mike", 20)

# # return statement
# def cube(num)
#   # will automatically return the last line if no 'return' word
#   num * num * num
#   5
# end
# puts cube(3) # 5
#
# def cube(num)
#   # will automatically return the last line if no 'return' word
#   return num * num * num
#   5
#   return 5
#   5
# end
# puts cube(3) # 27
#
# def cube(num)
#   # will automatically return the last line if no 'return' word
#   return num * num * num, 7
# end
# puts cube(3)[1] # 7

# # If statement
# isMale = true
# isTall = false
#
# if isMale and isTall
#   puts "You are a tall male"
# elsif isMale and !isTall
#   puts "You are a male but not tall"
# elsif !isMale and isTall
#   puts "You are not a male and you are tall"
# else
#   puts "You are not male or not tall or both"
# end

# # If statement using comparison
# def max(num1, num2, num3)
#   if num1 >= num2 and num1 >= num3
#     return num1
#   elsif num2 >= num1 and num2 >= num3
#     return num2
#   else
#     return num3
#   end
# end
# puts max(1, 3, -5)

# # Building a better calculator
# puts "Enter first number: "
# num1 = gets.chomp().to_f
# puts "Enter an operator: (+, -, *, /, %)"
# op = gets.chomp()
# puts "Enter second number: "
# num2 = gets.chomp().to_f
#
# if op == '+'
#   puts "The result is: "
#   puts (num1 + num2)
# elsif op == '-'
#   puts "The result is: "
#   puts (num1 - num2)
# elsif op == '*'
#   puts "The result is: "
#   puts (num1 * num2)
# elsif op == '/'
#   puts "The result is: "
#   puts (num1 / num2)
# elsif op == '%'
#   puts "The result is: "
#   puts (num1 % num2)
# else
#   puts ("Invalid operator")
# end

# # Case expressions (switch)
# def get_day_name(day)
#   day_name = ""
#   d = day.downcase()
#
#   case d
#   when "mon"
#     day_name = "Monday"
#   when "tue"
#     day_name = "Turesday"
#   when "wed"
#     day_name = "Wednesday"
#   when "thu"
#     day_name = "Thursday"
#   when "fri"
#     day_name = "Friday"
#   when "sat"
#     day_name = "Saturday"
#   when "sun"
#     day_name = "Sunday"
#   else # default
#     day_name = "Invalid abbreviation"
#   end
#
#   return day_name
# end
#
# puts get_day_name("mOn")
# puts get_day_name("mOnDaY")

# # While loops
# index = 1
# while index < 5
#   puts index
#   index += 1
# end

# # Build a guessing number
# secret_number = 7
# user_guess = nil
# guess_count = 3
# while user_guess != secret_number
#   puts ("You have " + guess_count.to_s + " time(s) left.")
#   puts "Enter a guess: "
#   user_guess = gets.chomp().to_f
#   guess_count -= 1
#
#   if user_guess > secret_number and guess_count != 0
#     puts "Make a smaller guess!"
#   elsif user_guess < secret_number and guess_count != 0
#     puts "Make a larger guess"
#   elsif user_guess == secret_number
#     puts "You Won!"
#   end
#
#   if guess_count <= 0
#     puts "You used all your chances!"
#     puts ("The secret number is " + secret_number.to_s)
#     return
#   end
# end

# # For loops
# friends = ["Kevin", "Karen", 'Oscar', "Angela", "Ruby"]
#
# for friend in friends
#   puts friend
# end
#
# friends.each do |friend|
#   puts friend
# end
#
# for index in 0..5 # inclusive on both side
#   puts index
# end
#
# 6.times do |index| # from 0 to i-1
#   puts index
# end

# # Exponent Methods for only positive exponent
# def pow(base, exp)
#   result = 1
#   exp.times do
#     result = result * base
#   end
#   return result
# end
#
# puts pow(2, 3) # 8
# puts pow(2, 0) # 1

# # Reading files
# # File.open(filepath, mode)
# File.open("employee.txt", "r") do |file|
# storing the file in file Variables
#   # # read first line
#   # puts file.readline()
#   # # read second line
#   # puts file.readline()
#   # # read char-wise
#   # puts file.readchar()
#   # puts file.readchar()
#
#   # # Second line
#   # puts file.readlines()[2]
#
#   for line in file.readlines()
#     puts line
#   end
# end
# # 2nd Way of open a file
# file = File.open("employee.txt", "r")
# puts file.read
# file.close()

# # Wrting Files, mode:
# # stackoverflow.com/questions/3682359/what-are-the-ruby-file-
# # open-modes-and-options
# File.open("employee.txt", "a") do |file| # only add information
#   # Adds Oscar, Accounting right after
#   file.write("\nOscar, Accounting")
# end
# # Overwrite the whole file
# File.open("employee.txt", "w") do |file|
#   file.write("Oscar, Accounting")
# end
#
# File.open("employee.txt", "r+") do |file|
#   file.readline()
#   # Replaces the second line first 9 chars to overriden
#   file.write("Overriden")
# end

# Heading Errors (USer rescue to handle error)
# num = 10/0 # ZeroDivisionError
# lucky_nums = [4, 8, 15, 16, 23, 42]
# lucky_nums["Dog"] # TypeError
#
# begin
#   num = 10/0
# rescue
#   puts "Division by zero error"
# end
#
# Not getting caught if only rescue ZeroDivisionError
# Specify specific error to catch
# begin
#   lucky_nums["Dog"]
#   num = 10/0
# rescue ZeroDivisionError
#   puts "Division by zero error"
# rescue TypeError => e # Store error message to e
#   puts "Wrong Type"
#   puts e
# end

# # Classes and Objects
# class Book
#   # attributes
#   attr_accessor :title, :author, :pages
# end
#
# book1 = Book.new()
# book1.title = "Harry Potter"
# book1.author = "JK Rowling"
# book1.pages = 400
#
# puts book1.title, book1.author, book1.pages
#
# book2 = Book.new()
# book2.title = "Lord of the Rings"
# book2.author = "Tolken"
# book2.pages = 500
#
# puts book2.title, book2.author, book2.pages

# # initialize Method
# class Book
#   # attributes
#   attr_accessor :title, :author, :pages
#   def initialize(title, author, pages)
#     puts ("Creating Book " + title + " ...")
#     @title = title
#     @author = author
#     @pages = pages
#   end
# end
#
# book1 = Book.new("Harry Potter", "JK Rowling", 400)
# puts book1.title, book1.author, book1.pages
# puts ""
# book2 = Book.new("Lord of the Rings", "Tolken", 500)
# puts book2.title, book2.author, book2.pages

# # Object/Instance methods
# class Student
#   # attributes
#   attr_accessor :name, :major, :gpa
#   def initialize(name, major, gpa)
#     puts ("Recording Student " + name + " ...")
#     @name = name
#     @major = major
#     @gpa = gpa
#   end
#
#   def has_honors
#     if @gpa >= 3.5
#       return true
#     else
#       return false
#     end
#   end
# end
#
# student1 = Student.new("Jim", "Business", 2.4)
# puts student1.has_honors
# puts ""
# student2 = Student.new("Pam", "CS", 3.6)
# puts student2.has_honors

# # Building a Quiz
#
# class Question
#   attr_accessor :prompt, :answer
#   def initialize(prompt, answer)
#     @prompt = prompt
#     @answer = answer
#   end
# end
#
# p1 = "What color are apples?\n(a)red\n(b)purple\n(c)orange"
# p2 = "What color are bananas?\n(a)pink\n(b)red\n(c)yellow"
# p3 = "What color are pears?\n(a)yellow\n(b)green\n(c)orange"
#
# questions = [
#   Question.new(p1, "a"),
#   Question.new(p2, "c"),
#   Question.new(p3, "b")
# ]
#
# def run_test(questions)
#   answer = ""
#   score = 0
#   for question in questions
#     puts question.prompt
#     answer = gets.chomp()
#     if answer == question.answer
#       score += 1
#     end
#   end
#   puts ("Your score: " + score.to_s + "/" + questions.length().to_s)
# end
#
# run_test(questions)

# # Inheritance
# #Superclass
# class Chef
#   def make_chicken
#     puts "The chef makes chicken"
#   end
#
#   def make_salad
#     puts "The chef makes salad"
#   end
#
#   def make_special
#     puts "The chef makes BBQ short ribs"
#   end
# end
# # Sub-class
# class ItalianChef < Chef
#   def make_special
#     puts "The chef makes pasta"
#   end
#
#   def make_pasta
#     puts "The chef makes pasta"
#   end
# end
#
# chef = Chef.new()
# chef.make_chicken()
# chef.make_special()
# # chef.make_pasta() not working
#
# italianchef = ItalianChef.new()
# italianchef.make_chicken()
# italianchef.make_special()
# italianchef.make_pasta()

# # Modules - self-created library
# module Tools
#   def sayHi(name)
#     puts "Hello, #{name}"
#   end
#   def sayBye(name)
#     puts "Bye, #{name}"
#   end
# end
#
# include Tools
# Tools.sayHi("Richard")
# Tools.sayBye("Richard")
#
# # use library in <filepath.rb> - like import
# # require_relative "filepath.rb"

# Interactive Ruby (irb)
# In terminal, type:
# irb
# 001> ...
# this is just like python ide
